lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+	{
+		FDCAN_CCCR(canport) |= FDCAN_CCCR_MON;
+	}
+	else
+	{
+		FDCAN_CCCR(canport) &= ~FDCAN_CCCR_MON;
+	}
+
+	if (rx_fifo_locked)
+	{
+		FDCAN_RXGFC(canport) &= ~(FDCAN_RXGFC_F1OM | FDCAN_RXGFC_F0OM);
+	}
+	else
+	{
+		FDCAN_RXGFC(canport) |= FDCAN_RXGFC_F1OM | FDCAN_RXGFC_F0OM;
+	}
+
+	if (loopback)
+	{
+		FDCAN_CCCR(canport) |= FDCAN_CCCR_TEST;
+		FDCAN_TEST(canport) |= FDCAN_TEST_LBCK;
+	}
+	else
+	{
+		FDCAN_CCCR(canport) &= ~FDCAN_CCCR_TEST;
+		/* FDCAN_TEST is automatically reset to default values by FDCAN 
+		 * at this point
+		 **/
+	}
+
+	return FDCAN_E_OK;
+}
+
+/** Enable FDCAN operation after FDCAN block has been set up.
+ * This function will disable FDCAN configuration effectively
+ * allowing FDCAN to sync up with the bus. After calling this function
+ * it is not possible to reconfigure amount of filter rules, yet
+ * it is possible to configure rules themselves.
+ * @param canport FDCAN block base address (use CANx defines)
+ * @returns FDCAN_E_OK on success, FDCAN_E_TIMEOUT if operation timed out.
+ * @Note If this function returns with timeout, it usually means that 
+ * FDCAN_clk is not set up properly.
+ */
+int fdcan_start(uint32_t canport)
+{
+	/* Error here usually means, that FDCAN_clk is not set up
+	 * correctly, or at all. This usually can't be seen above
+	 * when INIT is set to 1, because default value for INIT is
+	 * 1 as long as one has FDCAN_pclk configured properly. 
+	 **/
+	if (fdcan_cccr_init_cfg(canport, false) != 0)
+	{
+		return FDCAN_E_TIMEOUT;
+	}
+
+	return FDCAN_E_OK;
+}
+
+/** Configure amount of filters and initialize filtering block.
+ * This function allows configure global amount of filters present.
+ * FDCAN block will only ever check as many filters as this function configures.
+ * Function will also clear all filter blocks to zero values. This function
+ * can be only called after @see fdcan_init has already been called and
+ * @see fdcan_start has not been called yet as registers holding filter
+ * count are write-protected unless FDCAN block is in INIT mode.
+ * @param canport FDCAN block base address
+ * @param std_filt requested amount of standard ID filter rules (0-28)
+ * @param ext_filt requested amount of extended ID filter rules (0-8)
+ * @returns FDCAN_E_OK if filters were configured and initialized properly, otherwise
+ * return FDCAN_E_INVALID if FDCAN block base address is not recognized,
+ * FDCAN_E_OUTOFRANGE if requested amount of filters is too big,
+ * FDCAN_E_BUSY if device is already out of INIT mode.
+ */
+int fdcan_filter_init(uint32_t canport, uint8_t std_filt, uint8_t ext_filt)
+{
+	struct FDCAN_MessageRAM * ram = fdcan_get_msgram_addr(canport);
+
+	if ((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == 0)
+	{
+		return FDCAN_E_BUSY;
+	}
+
+	if (!ram)
+	{
+		return FDCAN_E_INVALID;
+	}
+
+	/* Only perform initialization of message RAM if there are
+	 * any filters required
+	 */
+	if (std_filt > 0)
+	{
+		if (std_filt > FDCAN_SFT_MAX_NR)
+		{
+			return FDCAN_E_OUTOFRANGE;
```

Comment:
same here, at a certain point, you should trust the users instead of trying to validate all inputs.